### PR TITLE
PE-75 kube autoscaler per ASG

### PR DESCRIPTION
**Pull request by sukenshah**
## Overview
The current version of `kube-aws-autoscaler` only allows fine-tuning parameters at the global level, which means all the worker pools are scaled up and down using the same criteria. This is good for a small number of similar profile worker pools but restricts us from setting up different scaling profiles/policies per worker pool. For instance, the scaling and buffer for interactive worker pools will be really different from the batch and core worker pools.

## Changes
The proposed solution is to fork the kube-aws-autoscaler repo and implement a new strategy where the autoscaler will use the following parameters from the tags of ASG of worker pools if present:

1. buffer memory %
2. buffer cpu %
3. buffer spare node

If the ASG doesn't have the tags then the autoscaler will use the global configured params, if they are missing then it'll fallback to hardcoded defaults.

## Unit testing
Added following unit tests for `calculate_buffer_per_auto_scaling_group` method:

test_calculate_buffer_per_auto_scaling_group_with_no_asg_tag
test_calculate_buffer_per_auto_scaling_group_with_cpu_override
test_calculate_buffer_per_auto_scaling_group_with_memory_override
test_calculate_buffer_per_auto_scaling_group_with_pods_override
test_calculate_buffer_per_auto_scaling_group_with_nodes_override
test_calculate_buffer_per_auto_scaling_group_with_all_override
test_calculate_buffer_per_auto_scaling_group_with_multiple_groups_and_partial_overrides

Some tests specifically setup using the buffer per asg for `calculate_required_auto_scaling_group_sizes` method:

test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_no_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_with_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_no_scale
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scale_down

## Integrated testing
Tested the changes by running `kube-aws-autoscaler` locally against the staging env in debug mode and with all fixed buffers set to zero and % buffer set to 10.

### 1. Core worker pool:
**ASG Settings**

Tag | Value
-- | --
kube-aws-autoscaler:buffer-cpu-percentage | 0
kube-aws-autoscaler:buffer-memory-percentage | 5
kube-aws-autoscaler:buffer-pods-percentage | 0
kube-aws-autoscaler:buffer-spare-nodes | 0

kube-aws-autoscaler-all-overrides

### 2.  Worker pool A:
**ASG Settings**

Tag | Value
-- | --
kube-aws-autoscaler:buffer-memory-percentage | 5

kube-aws-autoscaler-partial-overrides

Please note in this case the rest of the params are defaulted to 10%

### CODE REVIEW

File: kube_aws_autoscaler/main.py [hunk 1]
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+
+    groups = []
+    for key in nodes_by_asg_zone:
+        asg_name, zone = key
+        groups.append(asg_name)
+
+    if len(groups) > 0:
+        response = autoscaling.describe_auto_scaling_groups(AutoScalingGroupNames=groups)
+        for asg in response['AutoScalingGroups']:
+            if 'Tags' in asg.keys() and len(asg['Tags']) > 0:
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]:
```suggestion
            if 'Tags' in asg and asg['Tags']:
```

File: kube_aws_autoscaler/main.py [hunk 2]
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+                buffer_info = {}
+                for asg_tag in asg['Tags']:
+                    tag_key = asg_tag['Key']
+                    if tag_key in ASG_ALLOWED_TAGS.keys():
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]:
```suggestion
                    if tag_key in ASG_ALLOWED_TAGS:
```

File: kube_aws_autoscaler/main.py (hunk 1 above)
Comment (mousavian) [reply kube_aws_autoscaler/main.py]: WOW @ronny . TIL `suggestion`

File: kube_aws_autoscaler/main.py [hunk 3]
```diff
@@ -231,8 +232,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+                        buffer_info[ASG_ALLOWED_TAGS[tag_key]] = int(asg_tag['Value'])
+
+                if len(buffer_info.keys()) > 0:
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]:
```suggestion
                if buffer_info:
```

```
>>> bool({'a':1})
True
>>> bool({})
False
```

File: kube_aws_autoscaler/main.py [hunk 4]
```diff
@@ -46,6 +46,7 @@
 
 logger = logging.getLogger('autoscaler')
 
+ASG_ALLOWED_TAGS = {'buffer-memory-percentage': 'memory', 'buffer-cpu-percentage': 'cpu', 'buffer-pods-percentage': 'pods', 'buffer-spare-nodes': 'nodes'}
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]: maybe worth splitting these into lines to make it easier to read

Comment (msessa) [inline kube_aws_autoscaler/main.py]: Perhaps we should use a prefix on those tags to help identify their purpose at a glance.
Ex:
  aws-autoscaler-buffer-memory-percentage
  aws-autoscaler-buffer-cpu-percentage
  etc

File: kube_aws_autoscaler/main.py (hunk 2 above)
Comment (sukenshah) [reply kube_aws_autoscaler/main.py]: Fixed.

File: kube_aws_autoscaler/main.py (hunk 4 above)
Comment (sukenshah) [reply kube_aws_autoscaler/main.py]: Fixed.

File: kube_aws_autoscaler/main.py (hunk 3 above)
Comment (sukenshah) [reply kube_aws_autoscaler/main.py]: Fixed.

File: kube_aws_autoscaler/main.py (hunk 4 above)
Comment (sukenshah) [reply kube_aws_autoscaler/main.py]: Added `kube-aws-autoscaler:` as prefix.

File: kube_aws_autoscaler/main.py [hunk 5]
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+def calculate_buffer_per_auto_scaling_group(autoscaling, nodes_by_asg_zone: dict,
+                                            buffer_percentage: dict, buffer_spare_nodes: int=0):
+    buffer_per_asg = {}
+
+    groups = []
+    for key in nodes_by_asg_zone:
+        asg_name, zone = key
+        groups.append(asg_name)
+
+    if len(groups) > 0:
```
Comment (ronny) [inline kube_aws_autoscaler/main.py:249]:
```suggestion
    if groups:
```

File: kube_aws_autoscaler/main.py [hunk 6]
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+        response = autoscaling.describe_auto_scaling_groups(AutoScalingGroupNames=groups)
+        for asg in response['AutoScalingGroups']:
+            if 'Tags' in asg and asg['Tags']:
+                buffer_info = {}
+                for asg_tag in asg['Tags']:
+                    tag_key = asg_tag['Key']
+                    if tag_key in ASG_ALLOWED_TAGS:
+                        buffer_info[ASG_ALLOWED_TAGS[tag_key]] = int(asg_tag['Value'])
+
+                if buffer_info.keys():
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]:
```suggestion
                if buffer_info:
```

File: kube_aws_autoscaler/main.py [hunk 7]
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+                    # Fill the missing buffer resource with global
+                    if 'nodes' not in buffer_info.keys():
```
Comment (ronny) [inline kube_aws_autoscaler/main.py:261]:
```suggestion
                    if 'nodes' not in buffer_info:
```

File: kube_aws_autoscaler/main.py [hunk 8]
```diff
@@ -231,8 +237,42 @@ def slow_down_downscale(asg_sizes: dict, nodes_by_asg_zone: dict, scale_down_ste
...
+                        buffer_info['nodes'] = buffer_spare_nodes
+                    for buffer_resource in RESOURCES:
+                        if buffer_resource not in buffer_info.keys():
```
Comment (ronny) [inline kube_aws_autoscaler/main.py]:
```suggestion
                        if buffer_resource not in buffer_info:
```

File: kube_aws_autoscaler/main.py [hunk 9]
```diff
@@ -246,7 +286,15 @@ def calculate_required_auto_scaling_group_sizes(nodes_by_asg_zone: dict, usage_b
             # add requested resources from unassigned/pending pods
             for resource, val in pending.items():
                 requested[resource] += val
-        requested_with_buffer = apply_buffer(requested, buffer_percentage, buffer_fixed)
+
+        if asg_name in buffer_per_asg.keys():
```
Comment (ronny) [inline kube_aws_autoscaler/main.py:290]:
```suggestion
        if asg_name in buffer_per_asg:
```

**Comment by sukenshah** [comment]
Hello @ronny  @mousavian and @msessa I have added some unit tests for `calculate_buffer_per_auto_scaling_group `. See the PR description.

**Review by ronny — APPROVED** [review]
Tests are 💯. Awesome work!

**Comment by sukenshah** [comment]
Added some more following tests to cover ASG desired size.

test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_no_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scaleup_with_spare_node
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_no_scale
test_calculate_required_auto_scaling_group_sizes_with_asg_overrides_scale_down

File: kube_aws_autoscaler/main.py (hunk 6 above)
Comment (mousavian) [inline kube_aws_autoscaler/main.py]: this if statement `if buffer_info.keys():` is not necessary as you check again in lines 261 & 264 as well.

**Review by mousavian — APPROVED** [review]
great job @sukenshah

File: kube_aws_autoscaler/main.py (hunk 6 above)
Comment (sukenshah) [reply kube_aws_autoscaler/main.py]: The reason for checking this is to make sure we only add buffer info for an ASG with valid tags. It'll avoid creating buffer_info for ASG with tags but none with `kube-aws-autoscaler:`

**Pull request merged**